File: DoublyCL.c
#include<stdio.h>
#include<stdlib.h>

struct node{
    int data;
    struct node*next;
    struct node*prev;

};
typedef struct node NODE;
typedef struct node* PNODE;
typedef struct node** PPNODE;

void InsertFirst(PPNODE First,PPNODE Last,int No){

PNODE newn=NULL;
newn=(PNODE)malloc(sizeof(NODE));

newn->data=No;
newn->next=NULL;
newn->prev=NULL;

if((*First==NULL)&& (*Last==NULL)){
    
    *First=newn;
    *Last=newn;

}
else{

    newn->next=*First;
    (*First)->prev=newn;
    *First=newn;   
}
(*Last)->next = *First;
(*First)->prev = *Last;

}

void InsertLast(PPNODE First,PPNODE Last,int No){
PNODE newn=NULL;
PNODE temp=NULL;

newn=(PNODE)malloc(sizeof(NODE));

newn->data=No;
newn->next=NULL;
newn->prev=NULL;

if((*First==NULL) && (*Last==NULL)){
    
    *First=newn;
    *Last=newn;

}
else{

   (*Last)->next=newn;
   newn->next=*First;
   newn->prev=*Last;

   *Last=newn;


    
}
(*Last)->next = *First;
(*First)->prev = *Last;


}
void Display(PNODE First,PNODE Last){

if((First== NULL) && (Last == NULL))
    {
        printf("Linked List is emprty\n");
        return;
    }

    
    do
    {
        printf("%d\t",First->data);
        First = First -> next;
    }while(Last -> next != First);

}
int Count(PNODE First,PNODE Last){


int iCnt=0;

if((First== NULL) && (Last == NULL))
    {
        printf("Linked List is emprty\n");
        return 0;
    }
    do
    {
       iCnt++;

        First = First -> next;
    }while(Last -> next != First);

return iCnt;
}
void DeleteFirst(PPNODE First,PPNODE Last){


if((*First==NULL) && (*Last==NULL)){
  
   printf("unalbe to delete the LL is empty ");
   return;

}
else if(*First==*Last){
   
   free(*First);
   *First=NULL;
   *Last=NULL;
}
else{

    *First=(*First)->next;
    free((*Last)->next);
    (*First)->prev=*Last;
    (*Last)->next=*First;
}
}
void DeleteLast(PPNODE First,PPNODE Last){

 if((*First==NULL) && (*Last==NULL)){
  
   printf("unalbe to delete the LL is empty ");
   return;

}
else if(*First==*Last){
   
   free(*First);
   *First=NULL;
   *Last=NULL;

}
else{

    *Last=(*Last)->prev;
    free((*First)->prev);

    (*Last)->next=*First;
    (*First)->prev=*Last; 
}
}
void DeleteAtPos(PPNODE First,PPNODE Last,int iPos){



PNODE temp=NULL;


int i=0;
int iLength=0;

iLength=Count(*First,*Last);


if(iPos<1 || iPos>iLength+1){

    printf("invalide position : ");
    return;

}
if(iPos==1){

    DeleteFirst(First,Last);

}
else if(iPos==iLength+1){

    DeleteLast(First,Last);

}
else{
   
  temp=*First;
  for(i = 1; i < iPos -1; i++)
        {
            temp = temp -> next;
        }
        temp->next=temp->next->next;

        free(temp->next->prev);
        temp->next->prev=temp;
  
}
}
void InsertAtPos(PPNODE First,PPNODE Last,int No,int iPos){
int iLength=0;

PNODE temp=NULL;
PNODE newn=NULL;

int i=0;
iLength=Count(*First,*Last);


if(iPos<1 || iPos>iLength+1){

    printf("invalide position : ");
    return;

}
if(iPos==1){

    InsertFirst(First,Last,No);

}
else if(iPos==iLength+1){

    InsertLast(First,Last,No);

}
else{
   
   newn=(PNODE)malloc(sizeof(NODE));

   newn->data=No;
   newn->next=NULL;
   newn->prev=NULL;

   temp=*First;

    for(i=1;i<iPos-1;i++){

       temp=temp->next;

    }
    newn->next=temp->next;

    temp->next->prev=newn;

    temp->next = newn;
    newn->prev=temp;

}

}


int main(){

PNODE Head=NULL;
PNODE Tail=NULL;
int iRet=0;

InsertFirst(&Head,&Tail,51);
InsertFirst(&Head,&Tail,21);
InsertFirst(&Head,&Tail,11);


InsertLast(&Head,&Tail,101);
InsertLast(&Head,&Tail,111);
InsertLast(&Head,&Tail,121);


Display(Head,Tail);


iRet=Count(Head,Tail);
printf("Number of elements are : %d\n",iRet);






InsertAtPos(&Head,&Tail,105,5);

Display(Head,Tail);


iRet=Count(Head,Tail);
printf("Number of elements are : %d\n",iRet);

DeleteAtPos(&Head,&Tail,5);
Display(Head,Tail);


iRet=Count(Head,Tail);
printf("Number of elements are : %d\n",iRet);





    return 0;

}

File: DoublyCL.java

class DoubluCLCL{

    public static void main(String []Arg){
        int iRet=0;


        Doubly Aobj=new Doubly();

        Aobj.InsertFirst(11);
        Aobj.InsertFirst(21);
        Aobj.InsertFirst(51);

        Aobj.Display();
        iRet=Aobj.iCount();
        
        System.err.println("Numrber of elements are : "+iRet);

        Aobj.InsertLast(101);
        Aobj.InsertLast(111);
        Aobj.InsertLast(151);
        Aobj.InsertLast(201);
        Aobj.Display();
        iRet=Aobj.iCount();
        
        System.err.println("Numrber of elements are : "+iRet);

        Aobj.DeleteFirst();
        Aobj.Display();
        iRet=Aobj.iCount();
        
        System.err.println("Numrber of elements are : "+iRet);

       
        Aobj.Display();
        iRet=Aobj.iCount();
        
        System.err.println("Numrber of elements are : "+iRet);

        Aobj.InsertAtPos(2003,5);
        Aobj.Display();
        iRet=Aobj.iCount();
        
        System.err.println("Numrber of elements are : "+iRet);

        Aobj.DeleteAtPos(5);
        Aobj.Display();
        iRet=Aobj.iCount();
        
        System.err.println("Numrber of elements are : "+iRet);










    }
}class node{

    public int data;
    public node next;
    public node prev;

    public node(int iValue){
        next=null;
        prev=null;
        data=iValue;

    }

}

class Doubly{

    private node First;
    private node Last;
    private int iCount;

    public Doubly(){

        System.out.println("Inside constructor: ");
        First=null;
        Last=null;
        iCount=0;


    }
    public void InsertFirst(int iNo){

        node newn=null;

        newn=new node(iNo);

        if((First==null) || (Last==null)){

            First=newn;
            Last=newn;

        }
        else{

            newn.next=First;

            First.prev=newn;
            
            First=newn;
        }
       
        Last.next=First;
        First.prev=Last;

        iCount++;

    }
    public void InsertLast(int iNo){

        node newn=null;

        newn=new node(iNo);

        if(First==null || Last==null){

            First=newn;
            Last = newn;

        }
        else{

            newn.prev=Last;
            Last.next=newn;
            Last=newn;
        }
        First.prev=Last;
        Last.next=First;
        iCount++;

    }

    public void Display(){
       if((First==null ) && (Last==null)){

        System.out.println("LL is empty: ");
        return;

       }
       System.out.println("Elements of LL are : ");
       System.out.println("<=>");
    
       do{

        System.out.print("|"+First.data+"|<=>");
        First=First.next;
       } while(First !=Last.next);
       System.out.println();
    }


int iCount(){

    return iCount;

}

public void DeleteFirst(){

    if(First==null && Last==null){

        System.out.println("The LL is empty : ");
    }
     else if(First==Last){
        First=null;
        Last=null;

     }
     else{

        First=First.next;

     }
     Last.next = First;
     First.prev = Last;

     iCount--;
}
public void DeleteLast(){
    if(First==null && Last==null){

        System.out.println("The LL is empty : ");
    }
     else if(First==Last){
        First=null;
        Last=null;

     }
     else{

        Last=Last.prev;
        Last.next=First;

     }
     Last.next = First;
     First.prev = Last;

     iCount--;

}

public void InsertAtPos(int iNo,int iPos){

   node newn=null;
   node temp=null;

   newn=new node(iNo);

   if(iPos<1 || iPos>iCount+1){

           System.out.println("Invalid Position !!");
            return;

   }
   if(iPos==1){

    InsertFirst(iNo);

   }
   else if(iPos==iCount+1){

    InsertLast(iNo);

   }
   else{

    temp=First;
    int i=0;

    for(i=1;i<iPos-1;i++){
        temp=temp.next;

    }
    temp.next.prev=newn;
    newn.next=temp.next;
    newn.prev=temp;
    temp.next=newn;

    iCount++;

    
   }
}
public void DeleteAtPos(int iPos){
    
 
   node temp=null;

  

   if(iPos<1 || iPos>iCount+1){

           System.out.println("Invalid Position !!");
            return;

   }
   if(iPos==1){

   DeleteFirst();


   }
   else if(iPos==iCount+1){

    DeleteLast();

   }
   else{

    temp=First;
    int i=0;

    for(i=1;i<iPos-1;i++){
        temp=temp.next;

    }
    temp.next=temp.next.next;

    iCount--;
   
}
}
}
